[receiver/hostmetricsreceiver]: Add support for HugePages monitoring

sitole opened the pull request:
#### Description
Ability to collect Huge Mages memory metrics from host.

#### Link to tracking issue
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/42650

sitole committed:
Generated files

sitole committed:
Generated files

sitole committed:
Scrape huge pages metrics

sitole committed:
Do not use bytes

sitole committed:
Use bytes for huge pages page size

sitole committed:
Add changelog

paulojmdias commented on .chloggen/host-received-hugepages-metrics.yaml:
```diff
@@ -0,0 +1,27 @@
+change_type: enhancement
+
+# The name of the component, or a single word describing the area of concern, (e.g. receiver/filelog)
+component: receiver/hostmetricsreceiver
```
```suggestion
component: receiver/hostmetrics
```

thompson-tomo reviewed (changes requested):
Pr should be block until it is updated to follow latest semconv system design philosophy in particular naming https://github.com/open-telemetry/semantic-conventions/commit/a85e13b2df21d5a89967f21578a3ad165e7cae8f. This means that:
- metrics should be named `system.memory.linux.huge_pages.*` instead of `system.linux.memory.huge_pages.*`.

In addition IMO we should:
- consolidate `system.linux.memory.huge_pages.free`, `system.linux.memory.huge_pages.reserved` & `system.linux.memory.huge_pages.surplus` into a single metric `system.linux.memory.huge_pages.usage` with a state attribute to distinguish the different state just like we have with `system.memory.usage`.

- rename `system.linux.memory.huge_pages.total` metric to `system.linux.memory.huge_pages.limit` just like we have with `system.memory.limit`.

I would also suggest having a utilization metric so that we are mirroring the prior semconv art of system memory metrics.

sitole committed:
Rename and merge huge pages metrics, generate config

sitole committed:
Add huge pages utilization

thompson-tomo commented on receiver/hostmetricsreceiver/internal/scraper/memoryscraper/documentation.md:
```diff
@@ -60,6 +60,50 @@ Total bytes of memory available.
+
+| Name | Description | Values | Requirement Level |
+| ---- | ----------- | ------ | -------- |
+| state | Breakdown of memory usage by type. | Str: ``buffered``, ``cached``, ``inactive``, ``free``, ``slab_reclaimable``, ``slab_unreclaimable``, ``used``, ``reserved``, ``surplus`` | Recommended |
```
```suggestion
| system.memory.linux.hugepages.state | Breakdown of memory usage by type. | Str: ``buffered``, ``cached``, ``inactive``, ``free``, ``slab_reclaimable``, ``slab_unreclaimable``, ``used``, ``reserved``, ``surplus`` | Recommended |
```

thompson-tomo commented on receiver/hostmetricsreceiver/internal/scraper/memoryscraper/documentation.md:
```diff
@@ -60,6 +60,50 @@ Total bytes of memory available.
+
+| Name | Description | Values | Requirement Level |
+| ---- | ----------- | ------ | -------- |
+| state | Breakdown of memory usage by type. | Str: ``buffered``, ``cached``, ``inactive``, ``free``, ``slab_reclaimable``, ``slab_unreclaimable``, ``used``, ``reserved``, ``surplus`` | Recommended |
```
```suggestion
| system.memory.linux.hugepages.state | The Linux hugepages state. | Str: ``buffered``, ``cached``, ``inactive``, ``free``, ``used``, ``reserved``, ``surplus`` | Recommended |
```

Obviously filter values accordingly.

thompson-tomo reviewed (commented):
Definitions are looking alot better and are now consistent with existing definitions etc. Only suggestion would be to tweak this naming of the attribute to seperate it from existing memory state attribute just like what is planned for slab memory https://opentelemetry.io/docs/specs/semconv/system/system-metrics/#metric-systemlinuxmemoryslabusage

Reason for blocking has been addressed.

sitole committed:
Split system memory and linux hugepage memory state enum, added used pages metric as percentage and total number

paulojmdias reviewed (approved):
Just a documentation note.

sitole committed:
Fmt memory scraper metadata.yaml

sitole commented on receiver/hostmetricsreceiver/internal/scraper/memoryscraper/metadata.yaml:
```diff
@@ -14,7 +14,20 @@ attributes:
+  system.memory.linux.hugepages.state:
+    description: Breakdown of huge pages usage by state.
+    type: string
+    enum: [free, used, reserved, surplus]
```
Resolved, thx

paulojmdias reviewed (approved):
@sitole LGTM, but we need to wait for the code owners review.

@dmitryax @braydonk PTAL when you can 🙏

jakubno commented:
ping @dmitryax @braydonk

sitole committed:
Fix hugepages terminology in metadata.yaml

Replace 'huge pages' with 'hugepages' (one word) to match Linux kernel
documentation style and maintain consistency throughout the codebase.

Addresses review feedback from PR #44125.

sitole committed:
Change hugepages metric units from plural to singular

Update unit from '{pages}' to '{page}' for hugepages metrics to follow
OpenTelemetry convention where units are typically singular.

Addresses review feedback from PR #44125.

sitole committed:
Remove surplus from hugepages state enum

Remove 'surplus' from the state attribute enum to prevent double-counting.
According to Linux kernel docs, HugePages_Total = nr_hugepages + HugePages_Surp,
meaning surplus pages are already included in the total. Surplus pages can be
either free or used, so treating surplus as a mutually exclusive state would
cause the sum of states to exceed the total.

States are now: [free, used, reserved] which are mutually exclusive and sum
correctly to the total.

Addresses review feedback from PR #44125.

sitole committed:
Add separate surplus metric for hugepages

Add system.memory.linux.hugepages.surplus as a standalone gauge metric
instead of a state attribute. This provides visibility into hugepages
overcommit without breaking the additive property of the usage metric.

Since surplus pages can overlap with used/free states, exposing it
separately maintains the correct semantic that usage states should be
mutually exclusive and sum to the total.

Addresses review feedback from PR #44125.

sitole committed:
Update Go implementation to record surplus as separate metric

- Remove surplus from usage and utilization state-based recordings
- Add separate recording for surplus as a gauge metric
- Update comment to use 'hugepages' terminology
- Add explanatory comment about why surplus is separate

This matches the metadata.yaml changes where surplus was removed from
the state enum and added as a standalone metric.

Addresses review feedback from PR #44125.

sitole committed:
Update changelog to use hugepages terminology

Replace 'huge pages' with 'hugepages' in changelog to match the
terminology used throughout the codebase and Linux kernel documentation.

Addresses review feedback from PR #44125.

sitole committed:
Regenerate metadata files after hugepages metrics updates

Run mdatagen to regenerate all metadata-derived files:
- documentation.md: Updated with new terminology and surplus metric
- generated_config.go: Added surplus metric configuration
- generated_metrics.go: Added surplus metric recording method
- Test files and testdata updated accordingly

All generated files now reflect the metadata.yaml changes:
- 'hugepages' terminology
- Singular '{page}' unit
- Surplus as separate metric (not a state)
- State enum without surplus

Addresses review feedback from PR #44125.

sitole commented:
@braydonk changes applied, should be ready now.

braydonk reviewed (commented):
I had to do some more research to confirm this, so I apologize for not bringing it up sooner and causing another iteration loop.

I think that the `HugePages_Rsvd` needs to be excluded from the usage metric for a similar reason as `Surplus`, but the other way around. 

I can't find any documentation that explicitly states this is the case, but based on [`mm/hugetlb.c`](https://elixir.bootlin.com/linux/v6.18.6/source/mm/hugetlb.c#L2399) it seems that when hugepages are reserved, `h->resv_huge_pages` is incremented without the same amount being decremented from `h->free_huge_pages`. If I'm reading it right, then at the very least it confirms that the `HugePages_Free` count in `/proc/meminfo` (which are [directly retrieved from those `struct hstate` fields](https://elixir.bootlin.com/linux/v6.18.6/source/mm/hugetlb.c#L5235)) likely includes reserved hugepages. The `available_huge_pages` function in that file also leads me to believe that's the case, as the reserved pages are explicitly subtracted from free huge pages for that.

This means it should not be counted in the usage metric either. Perhaps similar to `system.memory.linux.hugepages.surplus`, it can be reported as a separate metric i.e. `system.memory.linux.hugepages.reserved`.

sitole committed:
Add separate reserved metric for hugepages

Move reserved hugepages from a usage state to a separate metric,
similar to how surplus is handled. This is necessary because reserved
pages are already counted in free pages (they represent a subset of
free pages that cannot be used for non-reserved allocations), so
including them as a usage state would break the semantic convention
that usage states must sum to the limit.

Also adds documentation explaining why reserved and surplus are
reported as separate metrics rather than usage states.

sitole committed:
Fix format

sitole commented:
Thanks for the feedback @rogercoll, @braydonk.

In c75a6d7a24c4e24872c845d9fe6001f972d696e8 reserved is removed from usage/utilization metrics and moved to its dedicated metrics, so now it should comply. I think even from our end, it's now easier to understand overcommitment in huge pages with the following metrics schema.

rogercoll reviewed (commented):
Code lgtm, just added a couple of nits. 

In addition, (if possible) I would suggest adding this metrics in [semantic-conventions](https://github.com/open-telemetry/semantic-conventions/) as it will be a requirement to mark them as stable.

sitole committed:
Changelog note

sitole committed:
Fix: remove double conversion to bytes

We were setting hugepage page size to bytes but we did it twice, once by
gopsutil and second time manually during writing metric value.

braydonk reviewed (approved):
Thank you for refining the design along with the feedback. LGTM now. I've opened https://github.com/open-telemetry/semantic-conventions/issues/3362 to track these metrics' contribution to Semantic Conventions.

paulojmdias reviewed (approved):
Thank you for the effort @sitole 🙏
